File: C++/710.random-pick-with-blacklist.cpp
/*
 * @lc app=leetcode id=710 lang=cpp
 *
 * [710] Random Pick with Blacklist
 *
 * https://leetcode.com/problems/random-pick-with-blacklist/description/
 *
 * algorithms
 * Hard (32.54%)
 * Likes:    244
 * Dislikes: 51
 * Total Accepted:    10.2K
 * Total Submissions: 31.3K
 * Testcase Example:  '["Solution", "pick", "pick", "pick"]\n[[1, []], [], [], []]'
 *
 * Given a blacklist B containing unique integers from [0, N), write a function
 * to return a uniform random integer from [0, N) which is NOT in B.
 * 
 * Optimize it such that it minimizes the call to system’s Math.random().
 * 
 * Note:
 * 
 * 
 * 1 <= N <= 1000000000
 * 0 <= B.length < min(100000, N)
 * [0, N) does NOT include N. See interval notation.
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: 
 * ["Solution","pick","pick","pick"]
 * [[1,[]],[],[],[]]
 * Output: [null,0,0,0]
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: 
 * ["Solution","pick","pick","pick"]
 * [[2,[]],[],[],[]]
 * Output: [null,1,1,1]
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: 
 * ["Solution","pick","pick","pick"]
 * [[3,[1]],[],[],[]]
 * Output: [null,0,0,2]
 * 
 * 
 * Example 4:
 * 
 * 
 * Input: 
 * ["Solution","pick","pick","pick"]
 * [[4,[2]],[],[],[]]
 * Output: [null,1,3,1]
 * 
 * 
 * Explanation of Input Syntax:
 * 
 * The input is two lists: the subroutines called and their arguments.
 * Solution's constructor has two arguments, N and the blacklist B. pick has no
 * arguments. Arguments are always wrapped with a list, even if there aren't
 * any.
 * 
 */
#include<vector>
#include<unordered_map>
#include<unordered_set>
using namespace std;

// @lc code=start
// class Solution {
// private:
//     int N;
//     unordered_set<int> blacklist;
// public:
//     Solution(int N, vector<int>& blacklist) {
//         this->N = N;
//         this->blacklist = {blacklist.begin(), blacklist.end()};
//     }
    
//     int pick() {
//         while (true) {
//             int num = rand() % N;
//             if (!blacklist.count(num)) return num;
//         }
//         return -1;
//     }
// };
class Solution {
private:
    int len;
    unordered_map<int, int> hash_map;
public:
    Solution(int N, vector<int>& blacklist) {
        len = N - blacklist.size();
        
        unordered_set<int> hash_set;
        for (int i=len; i<N; ++i) hash_set.insert(i);
        for (int num: blacklist) hash_set.erase(num);
        
        auto it = hash_set.begin();
        for (int num: blacklist) {
            if (num < len)
                hash_map[num] = *it++;
        }
        
    }
    
    int pick() {
        int num = rand() % len;
        return hash_map.count(num) ? hash_map[num] : num;
    }
};
/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(N, blacklist);
 * int param_1 = obj->pick();
 */
// @lc code=end

File: Python/710.random-pick-with-blacklist.py
#
# @lc app=leetcode id=710 lang=python3
#
# [710] Random Pick with Blacklist
#
# https://leetcode.com/problems/random-pick-with-blacklist/description/
#
# algorithms
# Hard (33.53%)
# Likes:    866
# Dislikes: 119
# Total Accepted:    43.1K
# Total Submissions: 128.4K
# Testcase Example:  '["Solution","pick","pick","pick","pick","pick","pick","pick"]\n' +
#   '[[7,[2,3,5]],[],[],[],[],[],[],[]]'
#
# You are given an integer n and an array of unique integers blacklist. Design
# an algorithm to pick a random integer in the range [0, n - 1] that is not in
# blacklist. Any integer that is in the mentioned range and not in blacklist
# should be equally likely to be returned.
# 
# Optimize your algorithm such that it minimizes the number of calls to the
# built-in random function of your language.
# 
# Implement the Solution class:
# 
# 
# Solution(int n, int[] blacklist) Initializes the object with the integer n
# and the blacklisted integers blacklist.
# int pick() Returns a random integer in the range [0, n - 1] and not in
# blacklist.
# 
# 
# 
# Example 1:
# 
# 
# Input
# ["Solution", "pick", "pick", "pick", "pick", "pick", "pick", "pick"]
# [[7, [2, 3, 5]], [], [], [], [], [], [], []]
# Output
# [null, 0, 4, 1, 6, 1, 0, 4]
# 
# Explanation
# Solution solution = new Solution(7, [2, 3, 5]);
# solution.pick(); // return 0, any integer from [0,1,4,6] should be ok. Note
# that for every call of pick,
# ⁠                // 0, 1, 4, and 6 must be equally likely to be returned
# (i.e., with probability 1/4).
# solution.pick(); // return 4
# solution.pick(); // return 1
# solution.pick(); // return 6
# solution.pick(); // return 1
# solution.pick(); // return 0
# solution.pick(); // return 4
# 
# 
# 
# Constraints:
# 
# 
# 1 <= n <= 10^9
# 0 <= blacklist.length <= min(10^5, n - 1)
# 0 <= blacklist[i] < n
# All the values of blacklist are unique.
# At most 2 * 10^4 calls will be made to pick.
# 
# 
#

# @lc code=start
class Solution:

    def __init__(self, n: int, blacklist: List[int]):
        self.num_white = n - len(blacklist)
        
        self.blacklist = set(blacklist)
        self.hmap = {}

        last = n - 1
        for black in blacklist:
            if black < self.num_white:
                while last in self.blacklist:
                    last -= 1
                self.hmap[black] = last
                last -= 1

    def pick(self) -> int:
        idx = random.randint(0, self.num_white - 1)
        if idx in self.hmap:
            return self.hmap[idx]
        return idx

# Your Solution object will be instantiated and called as such:
# obj = Solution(n, blacklist)
# param_1 = obj.pick()
# @lc code=end

